src/sponge/sponge.rs

```diff
@@ -0,0 +1,203 @@
+//! The `pad` module implements the Sponge's padding algorithm
+use super::pad::*;
+
+use dusk_bls12_381::Scalar;
+use dusk_plonk::constraint_system::{StandardComposer, Variable};
+use hades252::strategies::*;
+use hades252::WIDTH;
+
+/// The `hash` function takes an arbitrary number of Scalars and returns the
+/// hash, using the `Hades` ScalarStragegy
+pub fn sponge_hash(messages: &[Scalar]) -> Scalar {
+    let mut strategy = ScalarStrategy::new();
+
+    let mut words = pad(messages, WIDTH, Scalar::zero(), Scalar::one());
+    // If the words len is less than the Hades252 permutation `WIDTH` we directly
+    // call the permutation saving useless additions by zero.
+    if words.len() == WIDTH {
+        strategy.perm(&mut words);
+        return words[1];
+    }
+    // If the words len is bigger than the Hades252 permutation `WIDTH` then we
+    // need to collapse the padded limbs. See bottom of pag. 16 of
+    // https://eprint.iacr.org/2019/458.pdf
+    words
+        .chunks(WIDTH)
+        .fold(vec![Scalar::zero(); WIDTH], |mut inputs, values| {
+            let mut values = values.iter();
+            inputs
+                .iter_mut()
+                .for_each(|input| *input += values.next().unwrap());
+            strategy.perm(&mut inputs);
+            inputs
+        });
+    words[1]
+}
+
+/// The `hash` function takes an arbitrary number of plonk `Variable`s and returns the
+/// hash, using the `Hades` GadgetStragegy
+pub fn sponge_hash_gadget(composer: &mut StandardComposer, messages: &[Variable]) -> Variable {
+    let padder = composer.zero_var;
+    let eom = composer.add_input(Scalar::one());
+    let mut words = pad(messages, WIDTH, padder, eom);
+    // If the words len is less than the Hades252 permutation `WIDTH` we directly
+    // call the permutation saving useless additions by zero.
+    if words.len() == WIDTH {
+        let mut strategy = GadgetStrategy::new(composer);
+        strategy.perm(&mut words);
+        return words[1];
+    }
+    // If the words len is bigger than the Hades252 permutation `WIDTH` then we
+    // need to collapse the padded limbs. See bottom of pag. 16 of
+    // https://eprint.iacr.org/2019/458.pdf
+    words
+        .chunks(WIDTH)
+        .fold(vec![padder; WIDTH], |mut inputs, values| {
+            let mut values = values.iter();
+            inputs.iter_mut().for_each(|input| {
+                *input = composer.add(
+                    (Scalar::one(), *input),
+                    (Scalar::one(), *values.next().unwrap()),
+                    Scalar::zero(),
+                    Scalar::zero(),
+                )
+            });
+            let mut strategy = GadgetStrategy::new(composer);
```

Comment:
At this point you know better than me: since you always pass the same `composer`, do we need to create a new strategy all the time as well? Shouldn't we create it just once after the composer?

Reply:
Since it is mutably borrowed, I cannot create the strategy at the beggining of the function since then the `&mut Composer` would be already borrowed by the strategy and I will not be able to call it's functions or attributes.